### PR TITLE
Wildcard BUILDs

### PR DESCRIPTION
**Pull request by mikejholly**
This PR is my first pass on wildcard BUILD expansion. That is, builds in the form of `BUILD ./*+target`. This code will attempt to expand the BUILD into multiple commands if both the Earthfile & the target exist. The code will only explore directories 1 level beneath the referenced directory and globstar (`**`) is not yet supported.

Example: 

```
$ tree tests/wildcard-build 
tests/wildcard-build
├── bar
│   └── Earthfile
├── baz
│   └── Earthfile
├── foo
│   └── Earthfile
├── no-earthfile
└── no-target
    └── Earthfile
```

```
VERSION 0.7

test:
    FROM alpine
    BUILD ./wildcard-build/*+test
```

https://cloud.earthly.dev/builds/ec88fff1-d353-4045-a345-e3ce10cadff0

image

### CODE REVIEW

File: tests/wildcard-build.earth [hunk 1]
```diff
@@ -0,0 +1,9 @@
+VERSION 0.7
```
Comment (alexcb) [inline tests/wildcard-build.earth:1]: main now supports v0.8
```suggestion
VERSION 0.8
```

File: tests/Earthfile [hunk 2]
```diff
@@ -565,6 +566,11 @@ fail-invalid-artifact-test:
     DO +RUN_EARTHLY --earthfile=fail-invalid-artifact.earth --should_fail=true --target="--artifact +test/foo /tmp/stuff" \
         --output_contains="cannot save artifact +test/foo, since it does not exist"
 
+wildcard-build:
+    COPY --dir wildcard-build .
+    DO +RUN_EARTHLY --earthfile=wildcard-build.earth --target=+wildcard-build --output_does_not_contain="not-test run"
```
Comment (alexcb) [inline tests/Earthfile:571]: can you add some extra tests here to verify `earthly.output` (which is created by `DO +RUN_EARTHLY`) contains the expected "hello world" strings.

**Review by alexcb — COMMENTED** [review]
looking good! any thoughts on how hard it will be to support remote targets?

File: tests/Earthfile (hunk 2 above)
Comment (mikejholly) [reply tests/Earthfile:571]: Thanks. I forgot about that output file.

**Comment by vladaionescu** [comment]
> any thoughts on how hard it will be to support remote targets?

We are assuming that we don't support remote targets for this. (This is where I documented it: https://github.com/earthly/earthly/issues/3582)

But now that I'm thinking about this a bit more... I wonder what happens if you're invoking a remote target that in turn invokes a `BUILD ./*+something` within it. Will this work?

File: tests/Earthfile [hunk 3]
```diff
@@ -565,6 +566,15 @@ fail-invalid-artifact-test:
     DO +RUN_EARTHLY --earthfile=fail-invalid-artifact.earth --should_fail=true --target="--artifact +test/foo /tmp/stuff" \
         --output_contains="cannot save artifact +test/foo, since it does not exist"
 
+wildcard-build:
+    COPY --dir wildcard-build .
+    DO +RUN_EARTHLY --earthfile=wildcard-build.earth --target=+wildcard-build
+    RUN cat earthly.output | acbgrep 'hello world 1'
+    RUN cat earthly.output | acbgrep 'hello world 2'
+    RUN cat earthly.output | acbgrep 'hello world 3'
+    RUN cat earthly.output | (! acbgrep 'not-test run')
```
Comment (alexcb) [inline tests/Earthfile:575]:
```suggestion
    RUN cat earthly.output | acbgrep -v 'not-test run'
```

this will cause it to print an error stating the command failed, e.g.

```
alex@surfperch:~/gh/earthly/earthly$ echo "not-test run" | ./scripts/acbtest/acbgrep -v 'not-test run'
grep '-v' 'not-test run' failed
alex@surfperch:~/gh/earthly/earthly$ echo $?
1
```

where as using `(! ...` defeats the purpose of acbgrep (which is to print out a helpful message when it fails), 

for example, the output of a pass is confusing:

```
alex@surfperch:~/gh/earthly/earthly$ echo "blah" | (! ./scripts/acbtest/acbgrep 'not-test run')
blah
grep 'not-test run' failed
alex@surfperch:~/gh/earthly/earthly$ echo $?
0
```

File: tests/wildcard-build/foo/Earthfile [hunk 4]
```diff
@@ -0,0 +1,5 @@
+VERSION 0.7
```
Comment (alexcb) [inline tests/wildcard-build/foo/Earthfile:1]:
```suggestion
VERSION 0.8
```

**Comment by alexcb** [comment]
we should also add a feature flag for this, to make it more clear to users who are using an older version of earthly with this new feature.

**Comment by mikejholly** [comment]
Updated with feature flag & support for remote targets which use `BUILD` with wildcards.

Example:

```
earthly-dev github.com/earthly/test-remote/wildcard-build:882190030762e620e932e3fab1091bdd50c93131+wildcard-build
```

https://cloud.earthly.dev/builds/90ae164f-3bdf-4c25-a06b-b871b9ce805b

File: buildcontext/git.go [hunk 5]
```diff
@@ -66,6 +66,39 @@ type resolvedGitProject struct {
...
+	if !target.IsRemote() {
+		return nil, errors.Errorf("unexpected local reference %s", target.String())
+	}
+
+	rgp, _, subDir, err := gr.resolveGitProject(ctx, gwClient, platr, target)
+	if err != nil {
+		return nil, err
+	}
+
+	gitRef, err := llbutil.StateToRef(ctx, gwClient, rgp.state, false, platr.SubResolver(platutil.NativePlatform), nil)
```
Comment (mikejholly) [inline buildcontext/git.go:79]: I also tried to add `llb.Run(llb.Shlex("ls <path"))` here but I couldn't get it to work. The code here works well, but is limited to simple wildcard paths that end with `*` like `my/path/*`.

Comment (mikejholly) [reply buildcontext/git.go:79]: Here's the error. I found this surprising because the base image is `alpine/git:v2.43.0` and appears to include `ls`.

```
            internal *failed* | Repeating the failure error...
            internal *failed* | runc run failed: unable to start container process: exec: "/bin/sh": stat /bin/sh: no such file or directory
            internal *failed* | ERROR
            internal *failed* |       The internal command
            internal *failed* |           /bin/sh -c ls ./subdir/*
            internal *failed* |       did not complete successfully. Exit code 1
```

File: buildcontext/resolver.go [hunk 6]
```diff
@@ -83,6 +83,17 @@ func NewResolver(cleanCollection *cleanup.Collection, gitLookup *GitLookup, cons
 	}
 }
 
+// ExpandWildcard will expand a wildcard BUILD target in a remote Git
+// repository. The pattern is the path relative to the target path and should be
+// in the form 'my/path/*'
+func (r *Resolver) ExpandWildcard(ctx context.Context, gwClient gwclient.Client, platr *platutil.Resolver, target domain.Target, pattern string) ([]string, error) {
+	if !target.IsRemote() {
+		return nil, errors.Errorf("unexpected local reference %s", target.String())
```
Comment (vladaionescu) [inline buildcontext/resolver.go:91]: Should we just move the implementation for the local case over here too, for symmetry?

File: earthfile2llb/interpreter.go [hunk 7]
```diff
@@ -1239,6 +1248,79 @@ func (i *Interpreter) handleBuild(ctx context.Context, cmd spec.Command, async b
...
+		childTarget, err := domain.ParseTarget(childTargetName)
+		if err != nil {
+			return i.wrapError(err, cmd.SourceLocation, "failed to parse target %q", childTargetName)
+		}
+
+		data, _, _, err := i.converter.ResolveReference(ctx, childTarget)
+		if err != nil {
+			notExist := buildcontext.ErrEarthfileNotExist{}
+			if errors.As(err, &notExist) {
+				continue
```
Comment (vladaionescu) [inline earthfile2llb/interpreter.go:1289]: So if an Earthfile does not exist in a directory, it's not an error, right? Can you confirm that this works for both local and remote cases?

File: util/fileutil/fileutil.go [hunk 8]
```diff
@@ -65,3 +65,24 @@ func EnsureUserOwned(dir string, owner *user.User) error {
 		return os.Chown(path, uid, gid)
 	})
 }
+
+// GlobDirs will return any sub-directories which match the provided glob
+// pattern. Example: "/tmp/*" will return all sub-directories under "/tmp/".
+func GlobDirs(pattern string) ([]string, error) {
+	matches, err := filepath.Glob(pattern)
```
Comment (vladaionescu) [inline util/fileutil/fileutil.go:72]: The implementation for the local case seems much more rich than the remote version, since it permits more than just `*`. This can result in inconsistent behavior where Earthly builds something differently depending on how the target is referenced (can cause "works on my machine" type of issues).

Could we perhaps detect any unpermitted patterns here before passing to `filepath.Glob`? For example we could search for `[` or `?` and return an error in those cases.

(Maybe you're already checking for this somewhere else and I didn't see it)

File: buildcontext/git.go (hunk 5 above)
Comment (vladaionescu) [reply buildcontext/git.go:79]: Hmm.. not sure why that happens.

It may be worth trying to get that to work, since supporting patterns like `BUILD ./cmd-*+binary` would be useful.

File: buildcontext/resolver.go (hunk 6 above)
Comment (mikejholly) [reply buildcontext/resolver.go:91]: Sure, sounds good. I wasn't certain about where the new code belongs.

File: earthfile2llb/interpreter.go (hunk 7 above)
Comment (mikejholly) [reply earthfile2llb/interpreter.go:1289]: That's right. The path is ignored in that case. I'll add some more tests. There was one at one point, but I think I dropped it by mistake while refactoring.

File: util/fileutil/fileutil.go (hunk 8 above)
Comment (mikejholly) [reply util/fileutil/fileutil.go:72]: Yep, the code in `interpreter.go` currently limits support to the most basic case (e.g., `my/path/*`).

I can try to get `ls` working in the Git resolver to improve globbing support.

File: earthfile2llb/interpreter.go (hunk 7 above)
Comment (mikejholly) [reply earthfile2llb/interpreter.go:1289]: Oh, the directory was missing a `.gitkeep`. Thanks.

File: buildcontext/git.go (hunk 5 above)
Comment (alexcb) [reply buildcontext/git.go:79]: it might be a bit dirty (since it's no longer a git command), but maybe we could just add another line to https://github.com/earthly/earthly/blob/f51a0c1cbbbb44520d5343b373e66d8004f5ee75/buildcontext/git.go#L237

which would run a `find -name Earthfile > /dest/Earthfile-paths`, then a corresponding `gitMetaRef.ReadFile` which would return the list of Earthfiles in a new field under the `resolvedGitProject` struct.

Comment (vladaionescu) [reply buildcontext/git.go:79]: That's actually pretty reasonable IMO

Comment (mikejholly) [reply buildcontext/git.go:79]: Good idea. This would help remove the new code path in `gitResolver` & remove the need to run a custom `ls` or `find` command.

File: earthfile2llb/interpreter.go [hunk 9]
```diff
@@ -1239,6 +1248,79 @@ func (i *Interpreter) handleBuild(ctx context.Context, cmd spec.Command, async b
...
+
+	var matches []string
+
+	if isRemoteTarget {
+		matches, err = i.converter.ExpandRemoteWildcard(ctx, i.target, parsedTarget.GetLocalPath())
+	} else {
+		matches, err = fileutil.GlobDirs(parsedTarget.GetLocalPath())
+	}
+	if err != nil {
+		return i.wrapError(err, cmd.SourceLocation, "invalid BUILD wildcard pattern")
```
Comment (alexcb) [inline earthfile2llb/interpreter.go:1273]: the `"invalid BUILD wildcard pattern"` error message is confusing for cases where `ExpandWildcard` returns an error that's not related to the pattern (e.g. a network-related error)

Comment (mikejholly) [reply earthfile2llb/interpreter.go:1273]: Updated in upcoming commit. The language is now more generic.

**Comment by mikejholly** [comment]
Updated with improved Git target globbing. `filepath.Match` is useful here.

File: buildcontext/git.go [hunk 10]
```diff
@@ -310,6 +342,12 @@ func (gr *gitResolver) resolveGitProject(ctx context.Context, gwClient gwclient.
 		if err != nil {
 			return nil, errors.Wrap(err, "read git-refs")
 		}
+		earthfilePathsRaw, err := gitMetaRef.ReadFile(ctx, gwclient.ReadRequest{
```
Comment (vladaionescu) [inline buildcontext/git.go:345]: Just realizing that at some point it might be better to put all this metadata in a single file, to minimize the back-and-forth between earthly and buildkit with all these `ReadFile` requests. In any case, that can be a change that we make in a different PR.

File: buildcontext/git.go [hunk 11]
```diff
@@ -235,6 +266,7 @@ func (gr *gitResolver) resolveGitProject(ctx context.Context, gwClient gwclient.
 					"git log -1 --format=%ae >/dest/git-author || touch /dest/git-author ; " +
 					"git log -1 --format=%b >/dest/git-body || touch /dest/git-body ; " +
 					"git for-each-ref --contains HEAD --format '%(refname:lstrip=-1)' >/dest/git-refs || touch /dest/git-refs ; " +
+					"find -name Earthfile > /dest/Earthfile-paths || touch /dest/Earthfile-paths ; " +
```
Comment (vladaionescu) [inline buildcontext/git.go:269]: Maybe add `-type f`, to ensure that they're files and not dirs?

File: buildcontext/git.go [hunk 12]
```diff
@@ -63,7 +63,38 @@ type resolvedGitProject struct {
...
+		return nil, err
+	}
+
+	fullPattern := "./" + filepath.Join(subDir, pattern)
+
+	var matches []string
+
+	for _, path := range rgp.earthfilePaths {
+		path = strings.TrimSuffix(path, "/Earthfile")
+		ok, err := filepath.Match(fullPattern, path)
```
Comment (vladaionescu) [inline buildcontext/git.go:86]: Neat!

File: buildcontext/git.go [hunk 13]
```diff
@@ -63,7 +63,38 @@ type resolvedGitProject struct {
...
+	if !target.IsRemote() {
+		return nil, errors.Errorf("unexpected local reference %s", target.String())
+	}
+
+	rgp, _, subDir, err := gr.resolveGitProject(ctx, gwClient, platr, target)
+	if err != nil {
+		return nil, err
+	}
+
+	fullPattern := "./" + filepath.Join(subDir, pattern)
```
Comment (vladaionescu) [inline buildcontext/git.go:80]: Nit: Not sure if necessary, but should this path be cleaned? Like if it contains things like `././foo/bar/../../buz`?

File: earthfile2llb/converter.go [hunk 14]
```diff
@@ -1594,6 +1595,71 @@ func (c *Converter) Pipeline(ctx context.Context) error {
...
+		childTargetName := fmt.Sprintf("./%s+%s", match, parsedTarget.GetName())
+
+		childTarget, err := domain.ParseTarget(childTargetName)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to parse target %q", childTargetName)
+		}
+
+		data, _, _, err := c.ResolveReference(ctx, childTarget)
+		if err != nil {
+			notExist := buildcontext.ErrEarthfileNotExist{}
```
Comment (vladaionescu) [inline earthfile2llb/converter.go:1632]: Just realizing another edge case. What if the Earthfile does exist, but it does not contain the target we're looking for? Should we also skip that error?

**Review by vladaionescu — COMMENTED** [review]
Looks great! Some minor remaining remarks...

File: earthfile2llb/converter.go (hunk 14 above)
Comment (mikejholly) [reply earthfile2llb/converter.go:1632]: IIRC `ExpandWildcard` & `ResolveReference` do not verify that the target exists, but there's a check on L1639 which does. There are also tests for this scenario.

File: buildcontext/git.go (hunk 11 above)
Comment (mikejholly) [reply buildcontext/git.go:269]: Sure.

File: earthfile2llb/converter.go (hunk 14 above)
Comment (vladaionescu) [reply earthfile2llb/converter.go:1632]: Oh yeah - makes sense!

File: buildcontext/git.go (hunk 13 above)
Comment (mikejholly) [reply buildcontext/git.go:80]: `Join` should do most of the work but `Clean` will strip the `./` prefix. I'll add. 

https://go.dev/play/p/Q2GwzWroqVI

Comment (vladaionescu) [reply buildcontext/git.go]: Oh yeah, `Join` seems to be doing a clean of its own 👍

**Review by vladaionescu — APPROVED** [review]
🎉

File: buildcontext/git.go (hunk 13 above)
Comment (mikejholly) [reply buildcontext/git.go]: `Clean` actually strips off the leading `./`, which the following code expects.

Comment (vladaionescu) [reply buildcontext/git.go]: SG!

**Review by alexcb — APPROVED** [review]
nice work!
